rotkehlchen/balances/historical.py

```diff
@@ -99,63 +265,230 @@ def get_assets_amounts(
 
-            if direction == EventDirection.IN:
-                total_amount += event.balance.amount
-            else:  # EventDirection.OUT
-                total_amount -= event.balance.amount
+        Uses asset balances per day combined with historical prices to calculate
+        the total net worth in the user's profit currency for each day. Stops calculation
+        if negative balance would be created for any asset.
+
+        May raise:
+            - NotFoundError if no events exist in the specified time period.
 
-            amounts[ts_ms_to_sec(event.timestamp)] = total_amount
-            if total_amount < ZERO:
-                last_event_id = event.event_identifier
+        Returns:
+            - A mapping of timestamps to net worth in main currency for each day
+            - A list of (asset id, timestamp) tuples where price data was missing
+            - The event_identifier that caused a negative balance if any, else None
+        """
+        events, main_currency = self._get_events_and_currency(from_ts=from_ts, to_ts=to_ts)
+        if len(events) == 0:
+            raise NotFoundError(f'No historical data found within {from_ts=} and {to_ts=}.')
+
+        negative_balance_event = None
+        current_balances: dict[Asset, FVal] = defaultdict(FVal)
+        daily_balances: dict[Timestamp, dict[Asset, FVal]] = {}
+        current_day = timestamp_to_daystart_timestamp(events[0].timestamp)
+        for event in events:
+            if (day_ts := timestamp_to_daystart_timestamp(event.timestamp)) > current_day:
+                daily_balances[current_day] = current_balances.copy()
+                current_day = day_ts
+
+            if (negative_balance_id := event.update_balances(current_balances)) is not None:
+                negative_balance_event = negative_balance_id
                 break
 
-        return amounts, last_event_id
+        if current_day is not None and negative_balance_event is None:
+            daily_balances[current_day] = current_balances.copy()
+
+        net_worth_per_day: dict[Timestamp, FVal] = {}
+        missing_price_points: list[tuple[str, Timestamp]] = []
+        for day_ts, asset_balances in daily_balances.items():
+            assets_to_price = [
+                asset for asset, balance in asset_balances.items()
+                if balance != ZERO
+            ]
+            if len(assets_to_price) == 0:
+                continue
+
+            prices, missing = self._get_prices(
+                timestamp=day_ts,
+                assets=assets_to_price,
+                main_currency=main_currency,
+            )
+            missing_price_points.extend(missing)
+
+            day_total = sum(
+                balance * prices[asset]
+                for asset, balance in asset_balances.items()
+                if asset in prices and balance != ZERO
+            )
+            net_worth_per_day[day_ts] = FVal(day_total)
```

Comment:
Why do you need to do `FVal()` here? It should be FVal * Fval, so the result is already an FVal. If mypy can't find it that's mypy's problem. You should not call the constructor for no reason. You should figure out and fix the type error.